[1단계] HTTP 웹 서버 구현

seungy0 opened the pull request.

ljh131 commented on src/main/java/webserver/FileController.java:
```diff
@@ -0,0 +1,24 @@
+            return Response.file(FileIoUtils.loadFileFromClasspath(path.asFilePath()), css);
+        } catch (Exception e) {
+            e.printStackTrace();
+            return Response.error();
```
에러 스트링을 각 컨트롤러에서 만들지 않고,, 에러 객체를 정의해서 로직에서 던지고 메인 핸들러에서 이를 받아서 에러 응답을 내려주는식의 처리는 어떨까요?

dongjoon1251 replied:
RuntimeException을 던지도록 수정했습니다.

ljh131 commented on src/main/java/webserver/Router.java:
```diff
@@ -0,0 +1,28 @@
+import webserver.response.Response;
+
+public class Router {
+    private static final String USER_PATH = "/user";
```
라우터에서 컨트롤러 별로 라우팅되는 경로를 알고있기보다는, 컨트롤러 공통 인터페이스를 만들고 각 컨트롤러가 이를 구현해 라우터는 그 인터페이스로 라우팅 테이블을 알아오는 방식은 어떨까요? 조금 더 복잡하겠지만 확장성에서 더 좋을 것 같습니다.

dongjoon1251 replied:
다음 단계 미션에 포함되는 리팩토링이라 다음 단계에서 말씀해주신 부분 적용 고려해보겠습니다.

ljh131 commented on src/main/java/webserver/UserController.java:
```diff
@@ -0,0 +1,89 @@
+    private static final String USER_LIST_PATH = "/user/list";
+
+    private UserService userService;
+    private Handlebars handlebars;
```
템플릿엔진은 이 컨트롤러에서만 쓰는 것 보단 공용으로 쓸 수 있게 하면 좋을 것 같습니다.

dongjoon1251 replied:
TemplateUtils로 분리했습니다.

ljh131 commented on src/main/java/webserver/UserController.java:
```diff
@@ -0,0 +1,89 @@
+
+    private Response users() {
+        try {
+            return Response.file(handlebars.compile("user/list").apply(new UsersDto(userService.findAll())).getBytes(), ContentType.HTML);
```
그리고 여기서 템플릿을 읽어와 응답으로 바꾸는 부분을 직접 사용하기보단 컨트롤러의 수퍼 클래스 혹은 다른 곳에서 해주면 좋을 것 같습니다.

dongjoon1251 replied:
컨트롤러 관련 리팩토링이 다음 미션에 포함되어 있어서 다음 미션에서 진행하겠습니다.

ljh131 commented on src/main/java/webserver/request/RequestHeader.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public Integer getContentLength() {
+        return lines.stream()
```
헤더는 매번 파싱하기보다는 요청이 오면 1회 파싱해두고 map에 k,v 형태로 저장해두면 효율적일 것 같습니다.

ljh131 commented on src/main/java/webserver/request/RequestPath.java:
```diff
@@ -0,0 +1,57 @@
+package webserver.request;
+
+public class RequestPath {
+    private static final String STATIC_PATH_PREFIX = "./static";
```
요구사항에 path가 `./css/style` 이렇게 나와있긴 하던데요,

http path는 /로 시작해야 되는 것으로 알고있습니다. 참고삼아 커멘트 남깁니다.

```
     request-line   = method SP request-target SP HTTP-version CRLF
```
https://tools.ietf.org/html/rfc7230#section-3.1.1

```
     request-target = origin-form
                    / absolute-form
                    / authority-form
                    / asterisk-form

     origin-form    = absolute-path [ "?" query ]
```
https://tools.ietf.org/html/rfc7230#section-5.3.1

```
     absolute-path = 1*( "/" segment )
```
https://tools.ietf.org/html/rfc7230#section-2.7

segment 역시 링크에서 타고 가면 됩니다.

위 문법인 augmented bnf에 의하면 `/ 세그먼트`가 1회 이상 반복입니다.
https://tools.ietf.org/html/rfc5234#section-3.6

dongjoon1251 replied:
해당 부분은 URL의 path를 확인하는 부분이 아니라 파일 경로에 대한 Prefix라서 혼동을 드렸던 것 갔습니다.

ljh131 commented on src/main/java/webserver/request/RequestReader.java:
```diff
@@ -0,0 +1,51 @@
+            return IOUtils.readData(bufferedReader, contentLength);
+        } catch (IOException e) {
+            e.printStackTrace();
+            return "";
```
바디 읽는게 실패해서 ""를 리턴하면 로직에서 잘못 처리할 수도 있고 여러모로 그냥 에러를 리턴하는게 더 정확할 것 같습니다.

dongjoon1251 replied:
에러를 RuntimeException으로 던지도록 수정했습니다.

ljh131 commented on src/main/java/webserver/response/ResponseHeader.java:
```diff
@@ -0,0 +1,108 @@
+        }
+    }
+
+    private void setContentLength(int contentLength) {
```
content-length, content-type 들은 어차피 같은 http 헤더이니 http 헤더를 `setHeader(name, value)`로 설정해서 map에 k,v로 저장하면 더 generic할 것 같습니다. 그리고 setContentLength가 필요하다면, 그 메서드를 호출하면 되고요

ljh131 commented on src/main/java/webserver/response/ResponseHeader.java:
```diff
@@ -0,0 +1,108 @@
+    @Override
+    public String toString() {
+        return String.join("",
+                Optional.ofNullable(protocol).map(protocol -> String.format("%s %s \r\n", protocol.getMessage(), status.getMessage())).orElse(""),
```
이 부분이 응답으로 내려줄 스트링을 만드는 부분 같은데요, 뭔가 잘못된 상황에서 orElse로 ""를 내주게 되면 http 프로토콜상 적절치 않으니 에러를 내는게 좋을 것 같습니다.
좀 더 우아하게 처리하려면 미리 정의된 '500 internal server error'을 응답 스트링으로 만들어 이를 응답을 내주는 방식도 있을 것 같습니다.

dongjoon1251 replied:
HttpProtocol과 상태는 항상 있어야 하는 부분이라 없을 경우에 RuntimeException이 나도록 수정했습니다.

ljh131 commented on src/main/java/webserver/request/RequestHeader.java:
```diff
@@ -0,0 +1,60 @@
+
+    public boolean isLogined() {
+        return lines.stream()
+                .filter(line -> line.contains("Cookie"))
```
로그인 여부는 비지니스 로직에 가까운 부분일 것 같아서 응답 파싱하는 곳 보다는 다른 곳에 있으면 좋을 것 같습니다.

dongjoon1251 replied:
쿠키 포함 여부 판단하는 메소드로 수정하고 로그인 확인 메소드는 UserController로 이동했습니다.

ljh131 commented on src/main/java/webserver/request/RequestHeader.java:
```diff
@@ -0,0 +1,60 @@
+
+    public Integer getContentLength() {
+        return lines.stream()
+                .filter(line -> line.contains("Content-Length"))
```
`x-custom-header: Content-Length`같은 헤더가 있으면 문제될 것 같습니다.
그리고 헤더 이름은 case insensitive 하게 처리하는게 좋을 것 같습니다.

dongjoon1251 replied:
특정 키값에 대한 확인을 startWith로 변경해보겠습니다.

ljh131 commented on src/main/java/webserver/FileController.java:
```diff
@@ -0,0 +1,24 @@
+        return getFileResponse(path, ContentType.HTML);
+    }
+
+    private Response getFileResponse(RequestPath path, ContentType css) {
```
css라는 이름은 잘못 적으신 것 같습니다.

ljh131 reviewed (commented):
제가 리뷰어가 아닌데 실수로 리뷰 했습니다😄 
몇 가지 의견 남겼으니 참고해주시면 감사하겠습니다.

htrucci commented on src/main/java/model/UsersDto.java:
```diff
@@ -0,0 +1,19 @@
+public class UsersDto {
+    private final List<UserDto> users;
+
+    public UsersDto(List<User> users) {
+        this.users = IntStream.range(0, users.size())
+                .mapToObj(i -> new UserDto(i + 1, users.get(i)))
+                .collect(Collectors.toList());
+    }
```
순번을 주기 위한 로직으로 보이는데요.
DTO의 역할이 아니라고 생각됩니다. 
별도의 요구사항이 없으므로 view에서 처리하거나 service 혹은, database 로직에 있어야 할 것으로 생각됩니다.

dongjoon1251 replied:
리스트 출력부분 때문에 넣었었는데 다른 방식으로 구현하고 DTO의 인덱스는 제거했습니다.

htrucci commented on src/main/java/utils/RequestParser.java:
```diff
@@ -0,0 +1,63 @@
+
+    public static Map<String, String> getRequestParamsFromBody(String line) {
+
+        return Arrays.stream(line.split(REQUEST_PARAM_DELIMITER))
+                .map(attribute -> attribute.split(REQUEST_PARAM_KEY_VALUE_DELIMITER))
+                .collect(Collectors.toMap(param -> param[REQUEST_PARAM_KEY_INDEX], param -> decode(param[REQUEST_PARAM_VALUE_INDEX])));
+    }
```
value가 없으면 오류가 발생할 것 같습니다.

htrucci replied:
상수명이 길어서 조금 읽기 어려운 것 같기도 하고..

dongjoon1251 replied:
메소드로 key, value  가져오는 부분 분리하고 params length에 따라서 오류가 나게 하거나 빈값을 리턴하게 수정했습니다.

htrucci commented on src/main/java/webserver/UserController.java:
```diff
@@ -0,0 +1,89 @@
+import java.io.IOException;
+
+public class UserController {
+    private static final String USER_CREATE_PATH = "/user/create";
+    private static final String USER_LOGIN_PATH = "/user/login";
+    private static final String USER_LIST_PATH = "/user/list";
```
UserController에서는 path마다 /user 가 반복되므로 별도 상수로 두고 아래에서도 활용헀으면 좋겠습니다.

dongjoon1251 replied:
Router에서 USER_PATH로 만들어서 적용했습니다.

htrucci commented on src/main/java/webserver/UserController.java:
```diff
@@ -0,0 +1,89 @@
+        String path = request.getHeader().getPath().asUrl();
+        String method = request.getHeader().getMethod();
+        if (USER_CREATE_PATH.equals(path)) {
+            if ("GET".equals(method)) {
```
HttpMethod.GET.matches(method) 
위와 같이 구현해도 되고 HttpMethod를 별도 Enum으로 만들던가하여 String Equals를 isXX 등으로
변경하면 좋을 것 같습니다.

dongjoon1251 replied:
HttpMethod 사용해서 수정하겠습니다.

htrucci commented on src/main/java/webserver/UserController.java:
```diff
@@ -0,0 +1,89 @@
+
+    private Response login(Request request) {
+        if (userService.login(RequestParser.getRequestParamsFromBody(request.getBody()))) {
+            return Response.redirectWithCookie(request, "logined=true", "/index.html");
```
쿠키처리에 대한 부분은 어플리케이션 전반에 걸쳐 반복적으로 사용될 주제이므로
"logined=false" 값을 직접 넘기지 않고 별도의 유틸 등을 만들어 조작하면 더 깔끔해지지 않을까요?

dongjoon1251 replied:
Cookie 클래스로 분리했습니다.

htrucci commented on src/main/java/webserver/request/RequestReader.java:
```diff
@@ -0,0 +1,51 @@
+    private static List<String> readRequestHeader(BufferedReader bufferedReader) throws IOException {
+        List<String> lines = new ArrayList<>();
+        String line = bufferedReader.readLine();
+        while (!"".equals(line)) {
+            if (line == null) {
+                break;
+            }
```
아래와 같이 줄여도 문제가 없을까요?
`while (!ObjectUtils.isEmpty(line)) { 
}`

htrucci commented on src/test/java/service/UserServiceTest.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Test
+    public void saveUser() {
+        User expectedUser = new User("javajigi", "password", "%EB%B0%95%EC%9E%AC%EC%84%B1", "javajigi%40slipp.net");
```
@BeforeAll 어노테이션을 사용하여 반복적인 Database.addUser를 줄일 수 있을 것 같습니다

htrucci commented on src/main/java/webserver/UserController.java:
```diff
@@ -0,0 +1,89 @@
+    public UserController() {
+        userService = new UserService();
+        TemplateLoader loader = new ClassPathTemplateLoader();
+        loader.setPrefix("/templates");
+        loader.setSuffix(".html");
```
어플리케이션 전반적으로 사용되는 상수일 것 같습니다

dongjoon1251 replied:
TemplateUtils로 분리했습니다.

htrucci reviewed (changes requested):
안녕하세요~! 아직 이전 과정을 진행하고 있는지라,, 리뷰가 조금 늦었습니다 ^^;; 그 사이 ljh131 님이 좋은 코멘트를 많이 남겨주셨네요 ㅎㅎ 저도 몇가지 고민해본 의견 추가하여 Requeste changes로 요청드립니다. 확인 부탁드립니다~!

htrucci commented:
오잉 왜 코멘트가 없어졌지;;
@dongjoon1251 
머지했습니다. 보고 많이 배우게 되네요. 다음 단계도 화이팅입니다~!

seungy0 committed:
feat: GET 요청을 받을 때 요청한 페이지를 응답하는 기능 구현

seungy0 committed:
feat: 요청에 맞는 Content-Type 적용

stylesheet를 포함한 static 디렉토리 내부의 파일에 대해 적절하게 응답하도록 변경

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seungy0 committed:
feat: 회원가입을 위해 query string 파싱

요청 데이터를 담는 HttpRequest와 응답을 처리하는 HttpResponse 클래스 작성

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seungy0 committed:
feat: post 방식으로 요청을 보내는 회원가입 페이지 구현

- 헤더의 Content-Length만큼 Request에서 body를 읽어오도록 구현
- PostRequestHandler에서 요청에 대한 로직 처리

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

seungy0 committed:
feat: 회원가입 후 index.html로 redirect

Co-authored-by: Eungu-Kwon <kiwi9084@gmail.com>

shared-moon commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,62 @@
+        if (tokens == null || tokens.length < 2) {
+            throw new IllegalArgumentException("Invalid request line");
+        }
+        this.method = tokens[0];
```
각 인덱스 상수로 정의해서 무슨 의미인지 명확히 해주면 좋을 것 같아요 ~

shared-moon commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,62 @@
+
+    public HttpRequest(String requestLine, Map<String, String> header, String body) {
+        String[] tokens = requestLine.isEmpty() ? null : requestLine.split(" ");
+        if (tokens == null || tokens.length < 2) {
```
이런 부분은 규약에 관한 내용이라 조건만 보고는 의미를 파악하기가 어려울 것 같아요. 메서드로 추출해서 무슨 검증을 하는지 이름을 붙여주면 어떨까요 ?

shared-moon commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class HttpRequest {
```
RequestLine이랑 Headers를 분리해보는 건 어떨까요 ? 역할이 조금 달라 파싱부분까지 섞여있으니 좀 헷갈리는 것 같아요 ~

shared-moon replied:
요런 부분도 유닛테스트 추가되면 좋을 것 같아요 ~

shared-moon commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,62 @@
+    private final String method;
+    private final String path;
+    private final Map<String, String> header;
+    private final Map<String, String> query;
```
요건 약간 queryParams 가 더 맞지 않을까요 ?

shared-moon commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,62 @@
+        this.form = parseForm();
+    }
+
+    public boolean isMethod(String method) {
```
Method는 enum으로 관리해보는 건 어떨까요 ?

shared-moon commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,62 @@
+        Map<String, String> map = new HashMap<>();
+        for (String token : tokens) {
+            String[] keyValue = token.split("=");
+            map.put(keyValue[0], keyValue.length > 1 ? keyValue[1] : "");
```
혹시 모르니 key, value에 trim을 해주는 건 어떨까요 ?

그리고 value가 없을 때 빈값 넣어주는게 맞는걸까요 ? 지금 구현하신 내용을 보면 값이 빈게 아니라 아예 안들어오는거라 규약 위반인 것 같아서, 무시하거나 터뜨려야되지 않을까 싶어서요!

shared-moon commented on src/main/java/webserver/PostRequestHandler.java:
```diff
@@ -0,0 +1,36 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PostRequestHandler {
```
Post랑 Get으로 핸들러를 분리한 이유가 있을까요 ?
요청 파싱을 제외하면 그 뒤 비즈니스 로직은 Get이나 Post나 다를 게 없을 것 같아서요!

seungy0 replied:
감사합니다. 지금은 그렇지만 이 부분은 추후 로직이 바뀔게 명백해서 분리했습니다ㅎㅎ

shared-moon commented on src/main/java/webserver/PostRequestHandler.java:
```diff
@@ -0,0 +1,36 @@
+
+        HttpResponse httpResponse = new HttpResponse("404", "Not Found", null, null);
+        httpResponse.sendResponse(out);
+        DataBase.findAll().stream()
```
이건 혹시 여기 왜있는걸까요 ?

shared-moon commented on src/main/java/utils/IOUtils.java:
```diff
@@ -17,4 +20,30 @@ public static String readData(BufferedReader br, int contentLength) throws IOExc
+        List<String> lines = new ArrayList<>();
+        String line;
+
+        while (!(line = reader.readLine()).isEmpty()) {
```
line이 null인 경우는 어떻게 동작하나요 ?

seungy0 replied:
NPE가 발생할 것 같아 이에 대해 예외처리했습니다.

shared-moon commented on src/main/java/utils/IOUtils.java:
```diff
@@ -17,4 +20,30 @@ public static String readData(BufferedReader br, int contentLength) throws IOExc
+        return parseHeaders(lines);
+    }
+
+    private static Map<String, String> parseHeaders(List<String> lines) {
```
요거 괜히 반복문을 두 번 도는 것 같은데, 각 라인 돌 때 한번씩 수행하면 안될까요 ? 어차피 부르는 메서드도 헤더를 읽는 메서드같아서요 ~

shared-moon commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,55 +1,75 @@
+                headers.get("Content-Length") != null ? headers.get("Content-Length") : "0"));
+            HttpRequest httpRequest = new HttpRequest(requestLine, headers, body);
+
+            if (httpRequest.isMethod("GET") && responseResources(httpRequest.getPath(), out)) {
```
패스가 `/` 일 때도 index.html로 연결되도록 해주시면 좋을 것 같아요 ~ 테스트가 통과가 안되고 있어요!

shared-moon reviewed (approved):
승규님 안녕하세요!
1단계 미션 잘 진행 해 주셨네요 :)
코멘트를 좀 남겨 두었는데, 다음 미션 진행 전 확인 부탁드립니다.
고생 많으셨어요!
